docs/rfc/rfc-011-delete-gas.md

```diff
@@ -0,0 +1,140 @@
+of this nature is exactly what `CheckTx` exists for and there is no reason for the mempool
+to keep track of this data separately.
+
+Removing Gas from Tendermint and the mempool would allow for the 'mempool' to be a better
+abstraction: any transaction that arrived at `CheckTx` and passed the check will either be
+a candidate for a later block or evicted after a TTL is reached. All other transactions are 
+completely invalid and can be discarded forever.
+
+Removing gas will not be completely straightforward. It will mean ensuring that 
+equivalent functionality can be implemented outside of the mempool using the mempool's API.
+
+## Discussion
+
+This section catalogs the functionality that will need to exist within the Tendermint
+mempool to allow Gas to be removed and replaced by application-side bookkeeping.
+
+### Requirement: Provide Mempool Tx Sorting Mechanism
+
+Gas produces a market for inclusion in a block. On many networks, a [gas fee][cosmos-sdk-fees] is
+included in pending transactions. This fee indicates how much a user is willing to
+pay per unit of execution and the fees are distributed to validators.
+
+Validators wishing to extract higher gas fees are incentivized to include transactions 
+with the highest listed gas fees into each block. This produces a naturally ordering
+of the pending transactions. Applications wishing to implement a gas mechanism need
+to be able to order the transactions in the mempool. This can trivially be accomplished 
+by sorting transactions using the `priority` field available to applications as part of
+v0.35's `ResponseCheckTx` message.
+
+### Requirement: Allow Application-Defined Block Resizing
+
+When creating a block proposal, Tendermint pulls a set of possible transactions out of
+the mempool to include in the next block. Tendermint uses MaxGas to limit the set of transactions
+it pulls out of the mempool fetching a set of transactions whose sum is less than MaxGas.
+
+By removing gas tracking from Tendermint's mempool, Tendermint will need to provide a way for
+applications to determine an acceptable set of transactions to include in the block.
+
+This is what the new ABCI++ `PrepareProposal` method is useful for. Applications
+that wish to limit the contents of a block by an application-defined limit may
+do so by removing transactions from the proposal it is passed during `PrepareProposal`.
+
+Fully meeting the requirements of the current Gas system would mean adding functionality that is
+not currently provided the `PrepareProposal` method.
+Applications that wish to include _more_ transactions in the proposal than `PrepareProposal`
+initially passed to it cannot do so in the current design. [Issue 7750][issue-7750] has been filed to discuss
+and possibly solve for this.
+
+### Requirement: Handle Transaction Metadata
+
+Moving the gas mechanism into applications adds an additional piece of complexity
```

Comment:
The application already has to handle all the "interesting" bookkeeping for gas, since it has to assign gas values in the first place. The consensus node stores the value and does a crude kind of filtering with it, but as we've seen, even that isn't really sufficient—applications also want control over the ordering of transactions.

While it's true the application might have to keep a trivial data structure (e.g., a map from tx-hash to gas value, at a minimum), doing so would immediately give it a lot more flexibility at relatively low cost. Practically the app would probably want other metadata there too. That facility, combined with PrepareProposal, would give the application the control it wants over transaction selection, ordering, and composition, without the consensus node having to maintain anything.

If I understand correctly, the only question that leaves open is the fate of transactions that arrived in the mempool but were _not_ selected for the current block. 

Extending your idea of an `EvictTx` method a bit, I think a clean way to handle that would be for the node to call the application at the end of the height, saying: "Here are all the tx hashes I currently have buffered that you did not select. Tell me what to do with them."  The app replies with a yes/no for each hash:

1. Keep it, I may want it later (tx is retained for another height).
2. Drop it, it's no longer relevant (tx is dropped).

With that, there is no longer any need to track TTLs, gas values, or other metadata. 

There are still other concerns that may want separate treatment, but as far as control of the mempool, that would free the consensus node from the need to track non-opaque blobs of control data. As @tychoish has pointed out in other discussions, there may be some value in permitting _opaque_ metadata (e.g., to be gossiped), but we should try to ensure the node never has to make decisions based on such data (in contrast to gas today).

Reply:
> I think a clean way to handle that would be for the node to call the application at the end of the height, saying: "Here are all the tx hashes I currently have buffered that you did not select. Tell me what to do with them." The app replies with a yes/no for each hash:

I believe this is the idea behind re`CheckTx`. At the end of each height, Tendermint will run `CheckTx` on all the remaining transactions in its pool as a means of garbage collection.

Reply:
This conversation around transaction metadata held by the application runs somewhat parallel to the conversations I've been having with Sergio in regards to: https://github.com/tendermint/tendermint/issues/7663. 

Currently Tendermint provides transaction tracing via the RPC i.e. a user submits a tx and polls to see if it made it in a block and can request the corresponding merkle proof. Tendermint uses the sha256 hash as the transactions id, this also means that when a transaction is committed, all nodes are able to remove every transaction that was in the committed block from their mempool. As the block transactions are indexed, users can also use the hash to retrieve the transaction, the emitted events, and the respective proof.

The new ABCI++ calls allows for transactions to be modified and squashed together in `PrepareProposal`. This completely invalidates our previous process for tracking transactions because changing it will give a different hash and there's no way that other nodes who are not the proposer can know what the modified transaction originally corresponded to. One of the ideas was thus to force the application to provide a deterministic id that Tendermint can use for tracing and garbage collection. This would have to be done by storing it in the transaction itself. The Cosmos SDK already does this by having the address of the sender and a nonce in every transaction (to prevent the sender from submitting the same tx twice). In fact we explicitly state that all applications must implement some form of replay protection. 

I believe the Cosmos SDK similarly stores the gas in the tx bytes itself so I'm not sure if it will need to cache metadata but it conceivably could.

Reply:
Yes, the transaction ID problem is also something I've been wondering about. Having the application control ID assignment makes sense to me.

Something I'm still not really understanding is why the application needs to manipulate the mempool outside the context of PrepareProposal, and then perhaps once more at the end of the block to GC.